python/vineyard/contrib/ml/mxnet.py

```diff
@@ -17,3 +17,163 @@
+        meta['label_type_meta_'] = y.asnumpy().dtype.str
+    return client.create_metadata(meta)
+
+
+def mxnet_dataframe_builder(client, value, builder, **kw):
+    meta = ObjectMeta()
+    meta['typename'] = 'vineyard::DataFrame'
+    cols = kw.get('cols')
+    label = kw.get('label')
+    meta['label'] = to_json(label)
+    meta['columns_'] = to_json(cols)
+    meta['__values_-key-%d' % (len(cols) - 1)] = to_json(label)
+    meta.add_member('__values_-value-%d' % (len(cols) - 1), builder.run(client, value[1]))
+    for i in range(len(cols) - 1):
+        meta['__values_-key-%d' % i] = to_json(cols[i])
+        meta.add_member('__values_-value-%d' % i, builder.run(client, value[0][:, i]))
+    meta['__values_-size'] = len(cols)
+    meta['partition_index_row_'] = kw.get('partition_index', [0, 0])[0]
+    meta['partition_index_column_'] = kw.get('partition_index', [0, 0])[1]
+    meta['row_batch_index_'] = kw.get('row_batch_index', 0)
+    return client.create_metadata(meta)
+
+
+def mxnet_builder(client, value, builder, **kw):
+    typename = kw.get('typename')
+    if typename == 'Tensor':
+        return mxnet_tensor_builder(client, value, **kw)
+    elif typename == 'DataFrame':
+        return mxnet_dataframe_builder(client, value, builder, **kw)
+
+
+def mxnet_tensor_resolver(obj, resolver, **kw):
+    meta = obj.meta
+    data_shape = from_json(meta['data_shape_'])
+    label_shape = from_json(meta['label_shape_'])
+    data_name = meta['data_type_']
+    label_name = meta['label_type_']
+    data_type = normalize_dtype(data_name, meta.get('value_type_meta_', None))
+    label_type = normalize_dtype(label_name, meta.get('value_type_meta_', None))
+    data = np.frombuffer(memoryview(obj.member('buffer_data_')), dtype=data_type).reshape(data_shape)
+    label = np.frombuffer(memoryview(obj.member('buffer_label_')), dtype=label_type).reshape(label_shape)
+    return mx.gluon.data.ArrayDataset((data, label))
+
+
+def mxnet_dataframe_resolver(obj, resolver, **kw):
+    meta = obj.meta
+    if kw.get('label') == None:
+        label = from_json(meta['label'])
+    else:
+        label = kw.get('label')
+    columns = from_json(meta['columns_'])
+    if not columns:
+        return pd.DataFrame()
+    blocks = []
+    index_size = 0
+    for idx, name in enumerate(columns):
+        np_value = np.frombuffer(memoryview(obj.member('__values_-value-%d' % idx).member("buffer_")), dtype=np.float32)
+        index_size = len(np_value)
+        if BlockPlacement:
+            placement = BlockPlacement(slice(idx, idx + 1, 1))
+        else:
+            placement = slice(idx, idx + 1, 1)
+        values = np.expand_dims(np_value, 0)
+        blocks.append(Block(values, placement, ndim=2))
+    if 'index_' in meta:
+        index = resolver.run(obj.member('index_'))
+    else:
+        index = pd.RangeIndex(index_size)
+    df = pd.DataFrame(BlockManager(blocks, [pd.Index(columns), index]))
+    target = df[label].values.astype(np.float32)
```

Comment:
Could you please refer the resolvers in `tensorflow.py` for how to reuse `resolver_context()` to avoid such code duplication?

Reply:
Yes, I tried that but it gives reshape error.

Reply:
Is there any difference between the pandas dataframe you get here and the one with resolver_context()? If not, I didn't any reason why that dataframe will have a reshape error?

Reply:
There is a slight difference. Instead of directly using resolver.run, I have changed it to returning memoryview of the member "buffer"

Reply:
I don't think so. Inside resolver.run, it also uses np.from_buffer(memoryview(obj.member('buffer_')).

Reply:
Yes but here there is no reshaping

Reply:
But why the reshape can be removed? Without the reshape, the np_value will have a different shape than it is saved to vineyard. Then you still want to make a dataframe from this np_value? I don't think this will work. Correct me if I'm wrong.

Reply:
You post the error log when using resolver_context() here or on Slack. We can help check the log.

Reply:
Yes, Reshaping is important.